File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []

        for i, t in enumerate(nums):
            # if not the first index and the num to the left = current num -> go next
            if i > 0 and t == nums[i - 1]:
                continue

            l, r = i + 1, len(nums) - 1

            while l < r:
                threesum = t + nums[l] + nums[r]

                if threesum > 0:
                    r -= 1
                elif threesum < 0:
                    l += 1
                else:
                    res.append([t, nums[l], nums[r]])
                    l += 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1

        return res
                

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        //init a prefix array going front to back that keeps track of products so far
        vector<int> frontBack(nums.size(), 1);
        //do the same for a suffix array going back to front;
        vector<int> backFront(nums.size(), 1);
        //init the int vector that will be returned
        vector<int> products(nums.size());

        //loop through each index of nums starting w/ the second index
        for(int i = 1; i < nums.size(); i++)
        {
            //insert the product of the prev index of frontBack and the prev num
            frontBack[i] = frontBack[i - 1] * nums[i - 1];
        }
        //loop through each index of nums starting w/ the second to last index
        for(int i = nums.size() - 2; i >= 0; i--)
        {
            //insert the product of the next index of backFront and the next num
            backFront[i] = backFront[i + 1] * nums[i + 1];
        }
        //loop through each index of nums again this time using that index to calculate its product
        for(int i = 0; i < nums.size(); i++)
        {
            //insert the product as the product of the prefix and suffix indices
            products[i] = frontBack[i] * backFront[i];
        }

        //return products
        return products;
    }
};
